### PR TITLE
Make Batch implement a NodeAdder interface

### PR DESCRIPTION
**Pull request by hsanjuan**
Needed for: https://github.com/ipfs/go-unixfs/pull/41

### CODE REVIEW

File: batch.go [hunk 1]
```diff
@@ -124,6 +126,39 @@ func (t *Batch) Add(nd Node) error {
 	return t.err
 }
 
+// Remove removes a node from the DAGService. This is a non-batching
+// operation.
+func (t *Batch) Remove(ctx context.Context, c cid.Cid) error {
```
Comment (Stebalien) [inline batch.go]: This needs to remove the node from the queue (or flush the queue) and then remove the node.

File: batch.go [hunk 2]
```diff
@@ -124,6 +126,39 @@ func (t *Batch) Add(nd Node) error {
...
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// RemoveMany calls RemoveMany in the underlying DAGService.
+func (t *Batch) RemoveMany(ctx context.Context, nodes []cid.Cid) error {
+	return t.ds.RemoveMany(ctx, nodes)
```
Comment (Stebalien) [inline batch.go]: Same, needs to remove from the queue.

File: batch.go [hunk 3]
```diff
@@ -124,6 +126,39 @@ func (t *Batch) Add(nd Node) error {
...
+}
+
+// Get calls Get from the underlying DAGService.
+func (t *Batch) Get(ctx context.Context, c cid.Cid) (Node, error) {
+	return t.ds.Get(ctx, c)
```
Comment (Stebalien) [inline batch.go]: Same. Needs to get from the queue.

File: batch.go [hunk 4]
```diff
@@ -124,6 +126,39 @@ func (t *Batch) Add(nd Node) error {
...
+}
+
+// GetMany calls GetMany from the underlying DAGService.
+func (t *Batch) GetMany(ctx context.Context, nodes []cid.Cid) <-chan *NodeOption {
+	return t.ds.GetMany(ctx, nodes)
```
Comment (Stebalien) [inline batch.go]: ditto

**Comment by Stebalien** [comment]
Personally, I think separating out a `NodeWriter` interface is probably the right way to go.

**Comment by hsanjuan** [comment]
@Stebalien you mean `NodeWriter` being an interface with a single `Add` method, and using `NodeWriter` directly in places which only Add() (importers) instead of a full featured DAGService (and `Batch` would implement it) ?

**Comment by Stebalien** [comment]
```go
type NodeWriter interface {
    Add(context.Context, Node) error
    AddMany(context.Context, []Node) error
}
```

(we'd also need `Batch` to implement `AddMany` but that shouldn't be difficult).

We don't *have* to do it this way, I just don't want `batch.Add(ctx, x); batch.Get(ctx, x.Cid())` to fail.

(as names, `NodeWriter` is even *worse* than `NodeGetter`...)

**Comment by hsanjuan** [comment]
So it's one more interface to the mix (`NodeAdder` ?) vs. having a suboptimal DAGService implementation for methods we don't use right now.

Before, doing Get on the DAGService while adding stuff with the batch-importers would cause the same problems (Get on a recently added node could fail it was still uncommitted).

I'm leaning towards just doing `Commit` before non-add operations and documenting that `Batch` does it this way and if someone ever have an interest in improving it can always be done, but I'm honestly not sure what's best (both work, both are easy to implement).

**Comment by Stebalien** [comment]
> having a suboptimal DAGService implementation for methods we don't use right now.

No, having a *broken* DAGService implementation that sometimes works and sometimes doesn't. It doesn't matter if *we're* not using it, the issue is that it *looks* like it should work but won't.

> Before, doing Get on the DAGService while adding stuff with the batch-importers would cause the same problems (Get on a recently added node could fail it was still uncommitted).

Yes, but *that* makes perfect sense. The "batch" is effectively a transaction. However, if we're going to claim that `Batch` implements `DAGService`, it needs to actually *implement* `DAGService`.

> I'm leaning towards just doing Commit before non-add operations and documenting that Batch does it this way and if someone ever have an interest in improving it can always be done, but I'm honestly not sure what's best (both work, both are easy to implement).

We can, of course, do that. However, I really don't see the issue adding a new interface. It gives us a clean divide so applications can pick what they want to use:

* NodeGetter - read-only
* NodeAdder - append-only
* DAGService - RW dag service

**Comment by hsanjuan** [comment]
> No, having a broken DAGService...

I meant with the `Commit` before Get fix etc. It's suboptimal because it may make Gets really slow/defeat the purpose of batching.

> However, I really don't see the issue adding a new interface

Ok i'll do that then

**Comment by Stebalien** [comment]
> I meant with the Commit before Get fix etc. It's suboptimal because it may make Gets really slow/defeat the purpose of batching.

Ah, got it. Really, either way is fine with me as long as whatever we do is *correct*. However, this is really like the `io.Reader`/`io.Writer` split so I think it'll be fine.

File: batch.go [hunk 5]
```diff
@@ -124,6 +124,18 @@ func (t *Batch) Add(nd Node) error {
 	return t.err
 }
 
+// Add many calls Add for every given Node, thus batching and
+// commiting them as needed.
+func (t *Batch) AddMany(ctx context.Context, nodes []Node) error {
+	for _, n := range nodes {
+		err := t.Add(ctx, n)
```
Comment (hsanjuan) [inline batch.go]: Could potentially append all nodes directly but that might get us way over maxSize/maxNodes.

Comment (Stebalien) [reply batch.go]: That shouldn't really be an issue. Going over max nodes/size should just trigger a flush (we don't want to buffer too much in memory).

Maybe we should just:

1. Make `Add` call AddMany`.
2. Move the main logic to `AddMany` (and process all nodes at once).

**Comment by Stebalien** [comment]
@hsanjuan LGTM.

**Pull request merged**